jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,427 @@
+            if(!this.getTail().equals(path2.getTail())) {
+                return false;
+            }
+            if(!this.getHead().equals(path2.getHead())) {
+                return false;
+            }
+            if(!path2.getLabels().containsAll(this.getLabels())) {
+                return false;
+            }
+
+            return true;
+        }
+
+        /**
+         * Returns the start vertex of the path
+         *
+         * @return the start vertex of the path
+         */
+        protected V getHead() {
+            return vertices.get(0);
+        }
+
+        /**
+         * Returns the end vertex of the path
+         *
+         * @return the end vertex of the path
+         */
+        protected V getTail() {
+            return vertices.get(vertices.size() - 1);
+        }
+
+        /**
+         * Returns whether the path is empty, i.e. has no vertices
+         *
+         * @return whether the path is empty
+         */
+        protected boolean isEmpty() {
+            return vertices.isEmpty();
+        }
+
+        /**
+         * Returns an ordered list of the vertices of the path
+         *
+         * @return an ordered list of the vertices of the path
+         */
+        protected List<V> getVertices() {
+            return vertices;
+        }
+
+        /**
+         * Returns the labels of all vertices of the path
+         *
+         * @return the labels of all vertices of the path
+         */
+        protected Set<Integer> getLabels() {
+            return labels;
+        }
+
+        /**
+         * Returns the cost of the path
+         *
+         * @return the cost of the path
+         */
+        protected double getCost() {
+            return cost;
+        }
+
+        /**
+         * Returns a shallow copy of this labeled path instance. Vertices are not cloned.
+         *
+         * @return a shallow copy of this path.
+         *
+         * @throws RuntimeException in case the clone is not supported
+         *
+         * @see java.lang.Object#clone()
+         */
+        public LabeledPath<V> clone() {
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = (ArrayList<V>) this.vertices.clone();
+                newLabeledPath.labels = (HashSet<Integer>) this.labels.clone();
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    /**
+     * Implementation of an key for the path maps.
+     * It is used in AhujaOrlinSharmaCyclicExchangeLocalAugmentation to efficiently maintain the path sets in the calculation.
+     *
+     * @param <V> the vertex type
+     *
+     * @author Christoph Grüne
+     * @since June 7, 2018
+     */
+    private static class PathSetKey<V> {
```

Comment:
Don't you have to override equals/hashCode?  If the answer is "yes", then you might want to think about whether your unit tests (which are currently passing) are providing sufficient coverage.